File: internal/detect/compose.go
package detect

import (
	_ "embed"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"gopkg.in/yaml.v3"
)

//go:embed catalog.yaml
var catalogYAML []byte

// ServiceKind classifies a docker-compose service.
type ServiceKind string

const (
	KindApp     ServiceKind = "app"     // build: your code — runs native, not a shared service
	KindShared  ServiceKind = "shared"  // known backing image — a shared_service
	KindProxy   ServiceKind = "proxy"   // reverse proxy — Pomelo's dev-proxy replaces it
	KindUnknown ServiceKind = "unknown" // no image, no build
)

// ComposeService is one classified service from a docker-compose file.
type ComposeService struct {
	Name     string
	Kind     ServiceKind
	Image    string
	Type     string   // catalog type for a shared service ("custom" if unknown image)
	Strategy Strategy // isolation strategy for a shared service ("" if custom/unresolved)
	Clone    string   // per-branch seed method: template | dump-restore | none
	Ports    []int
	HasBuild bool
}

type catalogEntry struct {
	Type       string   `yaml:"type"`
	MatchImage []string `yaml:"match_image"`
	Strategy   Strategy `yaml:"strategy"`
	Ports      []int    `yaml:"ports"`
	Clone      string   `yaml:"clone"`
}

var proxyImages = []string{"nginx", "traefik", "caddy", "haproxy", "envoy"}

// Bare (un-namespaced) images that are backing services / dev tools, not the
// user's app — so a bare unknown image isn't blindly treated as a built app.
var bareBacking = []string{
	"adminer", "pgadmin", "phpmyadmin", "zookeeper", "consul", "vault", "etcd",
	"mosquitto", "eclipse-mosquitto", "prometheus", "grafana",
}

// ParseCompose finds a repo's docker-compose file and returns its classified
// services (backing → shared_service, build → app, reverse proxy → flagged).
func ParseCompose(repoPath string) []ComposeService {
	path := findComposeFile(repoPath)
	if path == "" {
		return nil
	}
	doc, err := readComposeFile(path)
	if err != nil {
		return nil
	}
	catalog := loadCatalog()
	dir := filepath.Dir(path)
	var out []ComposeService
	for _, name := range sortedKeys(doc.Services) {
		svc := doc.Services[name]
		image, hasBuild := resolveImage(svc, doc, dir, 0)
		cs := ComposeService{Name: name, Image: image, HasBuild: hasBuild}
		classify(&cs, catalog)
		out = append(out, cs)
	}
	return out
}

func classify(cs *ComposeService, catalog []catalogEntry) {
	if cs.HasBuild {
		cs.Kind = KindApp
		return
	}
	if cs.Image == "" {
		cs.Kind = KindUnknown
		return
	}
	img := strings.ToLower(cs.Image)
	for _, p := range proxyImages {
		if strings.Contains(img, p) {
			cs.Kind = KindProxy
			return
		}
	}
	for _, e := range catalog {
		for _, m := range e.MatchImage {
			if strings.Contains(img, strings.ToLower(m)) {
				cs.Kind = KindShared
				cs.Type = e.Type
				cs.Strategy = e.Strategy
				cs.Clone = e.Clone
				cs.Ports = e.Ports
				return
			}
		}
	}
	// Unknown image. A bare, un-namespaced name (e.g. "backend:latest") is almost
	// always a locally-built app image → treat as an app; a namespaced image
	// (acme/thing) is an unknown backing service the agent should classify.
	name := img
	if i := strings.IndexByte(name, ':'); i >= 0 {
		name = name[:i]
	}
	for _, b := range bareBacking {
		if name == b {
			cs.Kind = KindShared
			cs.Type = "custom"
			return
		}
	}
	if !strings.Contains(name, "/") {
		cs.Kind = KindApp
		return
	}
	cs.Kind = KindShared
	cs.Type = "custom"
}

type composeDoc struct {
	Services map[string]composeSvc `yaml:"services"`
}

type composeSvc struct {
	Image   string    `yaml:"image"`
	Build   yaml.Node `yaml:"build"`
	Extends yaml.Node `yaml:"extends"`
}

// resolveImage follows a service's `extends` chain (same-file, then file-based)
// to recover an image inherited from a base service.
func resolveImage(svc composeSvc, doc composeDoc, dir string, depth int) (string, bool) {
	if svc.Image != "" {
		return svc.Image, false
	}
	if svc.Build.Kind != 0 {
		return "", true
	}
	if depth > 6 || svc.Extends.Kind == 0 {
		return "", false
	}
	base, file := parseExtends(svc.Extends)
	if base == "" {
		return "", false
	}
	if file == "" {
		if b, ok := doc.Services[base]; ok {
			return resolveImage(b, doc, dir, depth+1)
		}
		return "", false
	}
	sub, err := readComposeFile(filepath.Join(dir, file))
	if err != nil {
		return "", false
	}
	if b, ok := sub.Services[base]; ok {
		return resolveImage(b, sub, filepath.Dir(filepath.Join(dir, file)), depth+1)
	}
	return "", false
}

func parseExtends(n yaml.Node) (service, file string) {
	switch n.Kind {
	case yaml.ScalarNode:
		return n.Value, ""
	case yaml.MappingNode:
		var m struct {
			Service string `yaml:"service"`
			File    string `yaml:"file"`
		}
		_ = n.Decode(&m)
		return m.Service, m.File
	}
	return "", ""
}

func findComposeFile(repoPath string) string {
	for _, name := range []string{"compose.yaml", "compose.yml", "docker-compose.yml", "docker-compose.yaml"} {
		p := filepath.Join(repoPath, name)
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}

func readComposeFile(path string) (composeDoc, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return composeDoc{}, err
	}
	var doc composeDoc
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return composeDoc{}, err
	}
	return doc, nil
}

func loadCatalog() []catalogEntry {
	var c []catalogEntry
	_ = yaml.Unmarshal(catalogYAML, &c)
	return c
}

func sortedKeys(m map[string]composeSvc) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

File: rust/crates/pom_detect/src/compose.rs
use std::path::Path;

use pom_config::yaml_node::{self, Node};

const CATALOG: &str = include_str!("../assets/catalog.yaml");
const PROXY_IMAGES: [&str; 5] = ["nginx", "traefik", "caddy", "haproxy", "envoy"];
/// Bare images that are backing services or dev tools, so an unknown bare image is not always the app.
const BARE_BACKING: [&str; 11] = [
    "adminer",
    "pgadmin",
    "phpmyadmin",
    "zookeeper",
    "consul",
    "vault",
    "etcd",
    "mosquitto",
    "eclipse-mosquitto",
    "prometheus",
    "grafana",
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum ServiceKind {
    /// `build:` - the user's own code, run natively.
    App,
    /// A known backing image - a shared service.
    Shared,
    /// A reverse proxy the dev proxy replaces.
    Proxy,
    /// No image and no build.
    Unknown,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ComposeService {
    pub name: String,
    pub kind: ServiceKind,
    pub image: String,
    /// The catalog type of a shared service ("custom" for an unknown image).
    pub kind_name: String,
    pub strategy: String,
    /// How a branch is seeded: template | dump-restore | none.
    pub clone: String,
    pub ports: Vec<u16>,
    pub has_build: bool,
}

struct CatalogEntry {
    kind_name: String,
    match_image: Vec<String>,
    strategy: String,
    ports: Vec<u16>,
    clone: String,
}

fn catalog() -> Vec<CatalogEntry> {
    let Ok(Some(root)) = yaml_node::parse(CATALOG) else {
        return Vec::new();
    };
    root.items()
        .unwrap_or_default()
        .iter()
        .map(|entry| CatalogEntry {
            kind_name: entry
                .get("type")
                .map(Node::text)
                .unwrap_or_default()
                .to_string(),
            match_image: entry
                .get("match_image")
                .and_then(Node::items)
                .unwrap_or_default()
                .iter()
                .map(|item| item.text().to_lowercase())
                .collect(),
            strategy: entry
                .get("strategy")
                .map(Node::text)
                .unwrap_or_default()
                .to_string(),
            ports: entry
                .get("ports")
                .and_then(Node::items)
                .unwrap_or_default()
                .iter()
                .filter_map(|port| port.text().parse().ok())
                .collect(),
            clone: entry
                .get("clone")
                .map(Node::text)
                .unwrap_or_default()
                .to_string(),
        })
        .collect()
}

fn classify(service: &mut ComposeService, catalog: &[CatalogEntry]) {
    if service.has_build {
        service.kind = ServiceKind::App;
        return;
    }
    if service.image.is_empty() {
        service.kind = ServiceKind::Unknown;
        return;
    }
    let image = service.image.to_lowercase();
    if PROXY_IMAGES.iter().any(|proxy| image.contains(proxy)) {
        service.kind = ServiceKind::Proxy;
        return;
    }
    if let Some(entry) = catalog.iter().find(|entry| {
        entry
            .match_image
            .iter()
            .any(|needle| image.contains(needle.as_str()))
    }) {
        service.kind = ServiceKind::Shared;
        service.kind_name = entry.kind_name.clone();
        service.strategy = entry.strategy.clone();
        service.clone = entry.clone.clone();
        service.ports = entry.ports.clone();
        return;
    }
    let name = image.split(':').next().unwrap_or_default();
    if BARE_BACKING.contains(&name) || name.contains('/') {
        service.kind = ServiceKind::Shared;
        service.kind_name = "custom".into();
    } else {
        // A bare unknown name (backend:latest) is nearly always an image built from the repo.
        service.kind = ServiceKind::App;
    }
}

fn read_compose(path: &Path) -> Option<Node> {
    yaml_node::parse(&std::fs::read_to_string(path).ok()?)
        .ok()
        .flatten()
}

fn service<'a>(document: &'a Node, name: &str) -> Option<&'a Node> {
    document.get("services")?.get(name)
}

/// The image a service runs, following `extends` (same file, then another file); `true` when it builds.
fn resolve_image(service_node: &Node, document: &Node, dir: &Path, depth: usize) -> (String, bool) {
    let image = service_node
        .get("image")
        .map(Node::text)
        .unwrap_or_default();
    if !image.is_empty() {
        return (image.to_string(), false);
    }
    if service_node.get("build").is_some() {
        return (String::new(), true);
    }
    let Some(extends) = service_node.get("extends").filter(|_| depth <= 6) else {
        return (String::new(), false);
    };
    let (base, file) = if extends.is_mapping() {
        (
            extends
                .get("service")
                .map(Node::text)
                .unwrap_or_default()
                .to_string(),
            extends
                .get("file")
                .map(Node::text)
                .unwrap_or_default()
                .to_string(),
        )
    } else {
        (extends.text().to_string(), String::new())
    };
    if base.is_empty() {
        return (String::new(), false);
    }
    if file.is_empty() {
        return match service(document, &base) {
            Some(parent) => resolve_image(parent, document, dir, depth + 1),
            None => (String::new(), false),
        };
    }
    let path = dir.join(&file);
    let Some(other) = read_compose(&path) else {
        return (String::new(), false);
    };
    match service(&other, &base) {
        Some(parent) => resolve_image(parent, &other, path.parent().unwrap_or(dir), depth + 1),
        None => (String::new(), false),
    }
}

/// The repo's docker-compose services, each classified; empty without a compose file.
pub fn parse_compose(repo: &Path) -> Vec<ComposeService> {
    let Some(path) = [
        "compose.yaml",
        "compose.yml",
        "docker-compose.yml",
        "docker-compose.yaml",
    ]
    .iter()
    .map(|name| repo.join(name))
    .find(|path| path.exists()) else {
        return Vec::new();
    };
    let Some(document) = read_compose(&path) else {
        return Vec::new();
    };
    let dir = path.parent().unwrap_or(repo);
    let catalog = catalog();
    let mut names: Vec<(String, Node)> = document
        .get("services")
        .and_then(Node::entries)
        .unwrap_or_default()
        .iter()
        .map(|(name, node)| (name.text().to_string(), node.clone()))
        .collect();
    names.sort_by(|a, b| a.0.cmp(&b.0));
    names
        .into_iter()
        .map(|(name, node)| {
            let (image, has_build) = resolve_image(&node, &document, dir, 0);
            let mut service = ComposeService {
                name,
                kind: ServiceKind::Unknown,
                image,
                kind_name: String::new(),
                strategy: String::new(),
                clone: String::new(),
                ports: Vec::new(),
                has_build,
            };
            classify(&mut service, &catalog);
            service
        })
        .collect()
}
